pkg/api/helpers_test.go

```diff
@@ -397,6 +398,275 @@ func TestMatchTaint(t *testing.T) {
 	}
 }
 
+func TestTolerationToleratesTaint(t *testing.T) {
+	genForgivenessSeconds := func(f int64) *int64 {
+		return &f
+	}
+
+	testCases := []struct {
+		description     string
+		toleration      Toleration
+		taint           Taint
+		expectTolerated bool
+	}{
+		{
+			description: "toleration and taint have the same key and effect, and operator is Exists, and taint has no value, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpExists,
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key and effect, and operator is Exists, and taint has some value, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpExists,
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key, effect and value, and operator is Equal, expect tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "bar",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: true,
+		},
+		{
+			description: "toleration and taint have the same key and effect, but different values, and operator is Equal, expect not tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "value1",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "value2",
+				Effect: TaintEffectNoSchedule,
+			},
+			expectTolerated: false,
+		},
+		{
+			description: "toleration and taint have the same key and value, but different effects, and operator is Equal, expect not tolerated",
+			toleration: Toleration{
+				Key:      "foo",
+				Operator: TolerationOpEqual,
+				Value:    "bar",
+				Effect:   TaintEffectNoSchedule,
+			},
+			taint: Taint{
+				Key:    "foo",
+				Value:  "bar",
+				Effect: TaintEffectNoExecute,
+			},
+			expectTolerated: false,
+		},
```

Comment:
So the ``TaintEffectNoExecute`` effect is the most repulsive? Saying to pods "Hey, you may be scheduled here, you may be admitted to start here. Good for you but I will not allow you to run here so all your effort is futile. You will be subject to de-resolution in a moment you start running. Unless you do not care I forbid running and it has consequences".

The way the Kubelet works (having a pod run asap) it does not make much sense to have any pod not tolerating ``TaintEffectNoExecute`` to be scheduled or admitted.

Does it make sense to assume ordering of taints? I.e. ``NoSchedule`` < ``NoAdmit`` < ``NoExecute``? With ``NoExecute`` set every pod is out of game the moment it steps to node. With ``NoAdmit`` set, it does not make sense to schedule a pod on the node as it will not be allowed to start at all. Which is basically the same as not allowing to be executed. Unless something good can happen after a pod is started but before executed. Though, static pods and pods controlled by daemon set ignore the scheduling phase, it still make sense to keep the pods running in a case a node becomes broken yet we want to keep logging and other monitoring stuff running.

Of course, the test is functional and has no effect on the eviction policy enforcement...